modules/mod-deep-learning/DeepModelManager.cpp

```diff
@@ -0,0 +1,554 @@
+   {
+      if (!this->IsInstalling(card))
+      {
+         Uninstall(card);
+         return;
+      }
+
+      if (!(httpCode == 200) && 
+          !(httpCode == 302))
+         Uninstall(card);
+          
+      
+      // let the caller handle this
+      handler(httpCode, body);
+
+      // get rid of the cached response
+      this->mResponseMap.erase(card->GetRepoID());
+   });
+
+   // download the model
+   try 
+   {
+      // save the metadata
+      wxLogDebug("saving modelcard for %s \n", card->GetRepoID());
+      card->SerializeToFile(
+         audacity::ToUTF8(wxFileName(GetRepoDir(card), "metadata.json").GetFullPath())
+      );
+
+      wxLogDebug("downloading model for %s \n",card->GetRepoID());
+
+      network_manager::ResponsePtr response = DownloadModel(card, progressHandler, installHandler);
+
+      // add response to a temporary map (in case of cancellation)
+      mResponseMap[card->GetRepoID()] = response;
+   }
+   catch (const char *msg)
+   {
+      wxLogError(msg);
+      return;
+   }
+}
+
+void DeepModelManager::Uninstall(ModelCardHolder card)
+{
+   bool success = true;
+   success &= wxRemoveFile(wxFileName(GetRepoDir(card), "model.pt").GetFullPath());
+   success &= wxRemoveFile(wxFileName(GetRepoDir(card), "metadata.json").GetFullPath());
+
+   if (!success)
+      throw ModelManagerException(XO("An error occurred while uninstalling the model."));
+   
+   wxRmDir(wxFileName(GetRepoDir(card)).GetFullPath());
+   
+}
+
+bool DeepModelManager::IsInstalling(ModelCardHolder card) const
+{
+   return mResponseMap.find(card->GetRepoID()) != mResponseMap.end();
+}
+
+void DeepModelManager::CancelInstall(ModelCardHolder card)
+{
+   if (!IsInstalling(card))
+    {
+      // should never really reach here (can't cancel an install that's not ongoing)
+      wxASSERT(false);
+      return;
+    }   
+   else
+   {
+      const std::string repoid = card->GetRepoID();
+      network_manager::ResponsePtr response = mResponseMap[repoid];
+      response->abort();
+      mResponseMap.erase(repoid);
+   }
+}
+
+ModelCardCollection DeepModelManager::GetCards(const std::string &effect_type) const
+{
+   ModelCardFilter filterId([&effect_type](ModelCardHolder card)
+   {
+      return card->effect_type() == effect_type;
+   });
+   return mCards.Filter(filterId);
+}
+
+void DeepModelManager::FetchModelCards(CardFetchedCallback onCardFetched, CardFetchProgressCallback onProgress)
+{
+   // add the card to our collection before passing to the callback
+   CardFetchedCallback onCardFetchedWrapper = [this, onCardFetched = std::move(onCardFetched)]
+   (bool success, ModelCardHolder card)
+   {
+      if (success)
+      {
+         // validate it and insert it into collection
+         try
+         {
+            std::lock_guard<std::mutex> guard(mCardMutex);
+               this->mCards.Insert(card);
+         }
+         catch (const InvalidModelCardDocument &e)
+         {
```

Comment:
The mutex locked above could be eliminated as I have suggested elsewhere, and `Insert` does not throw that exception type, so couldn't this simplify?